File: conformance/tests/oversize-bigint-shift/OversizeBigintShift.runar.zig
const runar = @import("runar");

// OversizeBigintShift -- a length-sensitive byte op whose operand is a bigint
// const too large for a JSON number (NEW-008). 9007199254740993 is 2^53 + 1,
// one past Number.MAX_SAFE_INTEGER, so the conformance runner cannot narrow it
// back to a JSON number and the checked-in expected-ir.json really does pin
// the "<n>n" STRING form that `runar-cli compile` writes to disk. On chain the
// operand is the 7-byte script number 01 00 00 00 00 00 20, OP_LSHIFT is
// length-preserving, and OP_INVERT flips all seven bytes.
pub const OversizeBigintShift = struct {
    pub const Contract = runar.StatefulSmartContract;

    out: i64 = 0,

    pub fn init(out: i64) OversizeBigintShift {
        return .{ .out = out };
    }

    pub fn shift(self: *OversizeBigintShift) void {
        self.out = ~(9007199254740993 << 8);
    }
};

File: conformance/tests/oversize-bigint-shift/OversizeBigintShift.runar.ts
import { StatefulSmartContract } from 'runar-lang';

/**
 * OversizeBigintShift -- a length-sensitive byte op whose operand is a bigint
 * const too large for a JSON number (NEW-008).
 *
 * `runar-cli compile` serialises every ANF bigint as a `"<n>n"` STRING, and
 * that artifact is what every SDK loads off disk. `conformance/runner` narrows
 * the SAFE-INTEGER ones back to JSON numbers when it stamps `expected-ir.json`,
 * so before this fixture only ONE of 71 goldens carried an `n`-string at all --
 * and that one has no byte ops. The result: nothing on disk ever fed a
 * `"<n>n"` const into `<<`, `>>`, `&`, `|`, `^` or `~`, and four of the seven
 * SDK ANF interpreters silently took the ByteString path for such an operand.
 *
 * `9007199254740993n` (2^53 + 1) is deliberately one past
 * `Number.MAX_SAFE_INTEGER`, so it survives the runner's narrowing and the
 * golden below really does pin the string form.
 *
 * On chain: `9007199254740993` is the 7-byte script number
 * `01 00 00 00 00 00 20`; `OP_LSHIFT 8` is LENGTH-PRESERVING, leaving
 * `00 00 00 00 00 20 00` (a NON-minimal 7-byte encoding of 2^45), and
 * `OP_INVERT` then flips all seven bytes to `ff ff ff ff ff df ff` =
 * -35993612646875135. An interpreter that re-minimises anywhere along that
 * chain disagrees with the deployed script and the continuation it builds is
 * unspendable.
 */
class OversizeBigintShift extends StatefulSmartContract {
  out: bigint;

  constructor(out: bigint) {
    super(out);
    this.out = out;
  }

  public shift(): void {
    this.out = ~(9007199254740993n << 8n);
  }
}
